nft_ingester/src/json_worker.rs

```diff
@@ -199,27 +205,28 @@ pub async fn run(json_downloader: Arc<JsonWorker>, rx: Receiver<()>) {
 
     let tasks_persister = TasksPersister::new(json_downloader.clone(), result_rx);
 
-    tasks_streamer.run(rx.resubscribe(), num_of_tasks, &mut workers_pool).await;
-    tasks_persister.run(rx.resubscribe(), &mut workers_pool).await;
+    tasks_streamer.run(cancellation_token.child_token(), num_of_tasks, &mut workers_pool);
+    tasks_persister.run(cancellation_token.child_token(), &mut workers_pool);
 
     for _ in 0..json_downloader.num_of_parallel_workers {
-        let cln_rx = rx.resubscribe();
         let json_downloader = json_downloader.clone();
         let tasks_rx = tasks_rx.clone();
         let result_tx = result_tx.clone();
+        let cancellation_token = cancellation_token.child_token();
 
         workers_pool.spawn(async move {
-            while cln_rx.is_empty() {
+            while !cancellation_token.is_cancelled() {
                 let mut locked_rx = tasks_rx.lock().await;
                 match locked_rx.try_recv() {
                     Ok(task) => {
                         drop(locked_rx);
 
                         let begin_processing = Instant::now();
 
-                        let response = json_downloader
-                            .download_file(task.metadata_url.clone(), CLIENT_TIMEOUT)
-                            .await;
+                        let response = tokio::select! {
```

Comment:
maybe, that `select` with `loop` can replace the `while`?

Reply:
will be reviewed in the next iterations of this PR